scene/gui/text_edit.cpp

```diff
@@ -1562,6 +1609,12 @@ void TextEdit::_notification(int p_what) {
 						}
 						char_ofs = 0;
 					}
+
+					HashMap<int, Variant> glyph_to_obj;
+					for (Variant k : TS->shaped_text_get_objects(rid)) {
+						glyph_to_obj[TS->shaped_text_get_object_glyph(rid, k)] = k;
+					}
```

Comment:
You can use `shaped_text_get_object_rect` and draw all of these at once, without mapping it to the glyphs.

Something like:
```c++
for (Variant k : TS->shaped_text_get_objects(rid)) {
	Rect2 col_rect = TS->shaped_text_get_object_rect(rid, k);
	col_rect.position += Vector2(char_margin + ofs_x, ofs_y);
	if (!clipped && (col_rect.position.x + char_margin) >= xmargin_beg && (col_rect.position.x + col_rect.size.x + char_margin) <= xmargin_end) {
		Array arr = k;
		if (arr.size() == 4 && arr[3].get_type() == Variant::Type::COLOR) {
			draw_rect(col_rect, Color(arr[3]));
		}
		draw_rect(col_rect, Color(1, 1, 1), false, 1);
	}
}
```